src/operation.rs

```diff
@@ -838,64 +852,77 @@ fn process_binary_operation(
-                )
-            }
+            let mut env = Environment::new();
+
+            let c1 = Closure {
+                body: RichTerm {
+                    term: t1,
+                    pos: pos1,
+                },
+                env: env1,
+            };
+            let c2 = Closure {
+                body: RichTerm {
+                    term: t2,
+                    pos: pos2,
+                },
+                env: env2,
+            };
 
-            let mut env: Environment = HashMap::new();
-            let res = match (*t1, *t2) {
-                (Term::Bool(b1), Term::Bool(b2)) => Term::Bool(b1 == b2),
-                (Term::Num(n1), Term::Num(n2)) => Term::Bool(n1 == n2),
-                (Term::Str(s1), Term::Str(s2)) => Term::Bool(s1 == s2),
-                (Term::Lbl(l1), Term::Lbl(l2)) => Term::Bool(l1 == l2),
-                (Term::Sym(s1), Term::Sym(s2)) => Term::Bool(s1 == s2),
-                (Term::Record(m1), Term::Record(m2)) => {
-                    let (left, center, right) = merge::hashmap::split(m1, m2);
+            match eq(&mut env, c1, c2) {
+                EqResult::Bool(b) => Ok(Closure::atomic_closure(Term::Bool(b).into())),
+                EqResult::Eqs(fst, snd, eqs) => Ok(Closure {
+                    body: RichTerm::new(Term::Op2(BinaryOp::MultiEq(eqs), fst, snd), pos_op),
+                    env,
+                }),
+            }
+        }
+        BinaryOp::MultiEq(eqs) => {
+            let mut env = Environment::new();
 
-                    if !left.is_empty() || !right.is_empty() {
-                        Term::Bool(false)
-                    } else {
-                        eq_all(
-                            center.into_iter().map(|(_, (t1, t2))| (t1, t2)),
-                            &env1,
-                            &env2,
-                            &mut env,
-                        )
-                    }
-                }
-                (Term::List(l1), Term::List(l2)) if l1.len() == l2.len() => {
-                    eq_all(l1.into_iter().zip(l2.into_iter()), &env1, &env2, &mut env)
-                }
-                (_, _) => Term::Bool(false),
+            let c1 = Closure {
+                body: RichTerm {
+                    term: t1,
+                    pos: pos1,
+                },
+                env: env1,
+            };
+            let c2 = Closure {
+                body: RichTerm {
+                    term: t2,
+                    pos: pos2,
+                },
+                env: env2,
             };
 
-            Ok(Closure {
-                body: res.into(),
-                env,
-            })
+            let mut eqs: Vec<_> = eqs
+                .into_iter()
+                .map(|(c1, c2)| {
+                    (
+                        c1.body.closurize(&mut env, c1.env),
+                        c2.body.closurize(&mut env, c2.env),
+                    )
+                })
+                .collect();
```

Comment:
I'm confused. Isn't everything properly closurized by `eq`, by assumption?

Reply:
Indeed, `eq` already closurizes everything.

This confusing trickery is necessary because of the (old) way operators are closurized. Operator are parametrized by a `CapturedTerm` type which represents the type of captured terms, that is stored terms other than their argument (typically, for `MultiEq`, the list of equalities to process). This parameter is initially instantiated with `RichTerm`, and when the abstract machine evaluates an application of an operator to some arguments, it maps these potential captured terms  to `Closure`s with an appropriate environment. So at this point, `eqs` is of type `Vec<(Closure, Closure)>`[1].

This is really just a representation/typing problem: at the beginning of this snippet, eqs is `Vec<(Closure, Closure)>`, while what you need are `RichTerm`, so you have to closurize `ci`s back to `RichTerm`s. This adds useless overhead, and this ought to be fixed, by making operators always hold `RichTerm` and use the standard `closurize()`[2] method instead.

-----------------------------------
[1]: `Closure` is basically a product `(RichTerm, Environment)`\
[2]: `closurize()` acts on a `RichTerm` and returns a `RichTerm`: it generates a fresh variable and bind it to a closure with the term in some given environment, then return this variable.